strawberryfields/utils/post_processing.py

```diff
@@ -0,0 +1,341 @@
+            of (shots, modes)
+
+    Returns:
+        array: array of dimension :math:`\underbrace{D\times D\times D\cdots\times D}_{\text{num modes}}`
+            containing the Fock state probabilities, where :math:`D` is the Fock basis cutoff truncation
+    """
+    _check_samples(photon_number_samples)
+
+    num_modes = photon_number_samples.shape[1]
+
+    shots = photon_number_samples.shape[0]
+    num_modes = photon_number_samples.shape[1]
+    max_val = np.max(photon_number_samples)
+    bins = [list(range(max_val + 2)) for i in range(num_modes)]
+    H, _ = np.histogramdd(photon_number_samples, bins=bins)
+    probabilities = H / shots
+    return probabilities
+
+
+def _product_for_modes(samples, modes=None):
+    """Getting the product of samples across modes.
+
+    Args:
+        samples (ndarray): the photon number samples with a shape
+            of (shots, modes)
+        modes (Sequence): a flat sequence containing indices of modes to get
+            the expectation value for
+
+    Returns:
+        ndarray: product of the samples across modes
+    """
+    modes = np.array(modes)
+    selected_samples = samples[:, modes]
+    return np.prod(selected_samples, axis=1)
+
+
+def _check_samples(samples):
+    """Validation check for the input samples.
+
+    Checks include data types checks and dimension of the input samples.
+
+    Args:
+        samples (ndarray): the photon number samples with a shape of (shots,
+            modes)
+    """
+    if not isinstance(samples, np.ndarray) or samples.ndim != 2:
+        raise Exception("Samples needs to be represented as a two dimensional numpy array.")
+
+
+def _check_modes(samples, modes):
+    """Validation checks for the input modes.
+
+    Checks include data types checks, checks for the dimension of the inputs
+    and the validity of the modes specified.
+
+    Args:
+        samples (ndarray): the photon number samples with a shape
+            of (shots, modes)
+        modes (Sequence): the input modes to get the expectation value for, a
+            flattened sequence
+    """
+    num_modes = samples.shape[1]
+    flattened_sequence_indices_msg = (
+        "The input modes need to be specified as a flattened sequence of indices!"
+    )
+
+    modes = np.array(modes)
+
+    # Checking if modes is a valid flattened sequence of indices
+    try:
+        non_index_modes = modes[(np.mod(modes, 1) != 0) | (modes < 0)]
+    except Exception:
+        raise Exception(flattened_sequence_indices_msg)
+
+    if modes.ndim != 1 or non_index_modes.size > 0 or len(modes) == 0:
+        raise Exception(flattened_sequence_indices_msg)
+
+    # Checking if valid modes were specified
+    largest_valid_index = num_modes - 1
+    out_of_bounds_modes = modes[modes > largest_valid_index]
+    if out_of_bounds_modes.size > 0:
+        raise Exception(
+            "Cannot specify mode indices {} for a {} mode system!".format(
```

Comment:
```suggestion
            "Cannot specify mode index {} for a {} mode system.".format(
```

Reply:
Note that f-strings are now allowed in SF and PL if you want to use them :)

Reply:
Thanks added them! (kept the plural form as a list can be the first argument)